syntax = "proto3";

package filemanagement;

service FileService {
  rpc CreateFile(CreateFileRequest) returns (CreateFileResponse);
  rpc ReadFile(ReadFileRequest) returns (ReadFileResponse);
  rpc WriteFile(WriteFileRequest) returns (WriteFileResponse);
  rpc DeleteFile(DeleteFileRequest) returns (DeleteFileResponse);
  rpc ListFiles(ListFilesRequest) returns (ListFilesResponse);
  rpc CreateDirectory(CreateDirectoryRequest) returns (CreateDirectoryResponse);
  rpc GetFileInfo(GetFileInfoRequest) returns (GetFileInfoResponse);
  rpc UploadFile(stream UploadFileRequest) returns (UploadFileResponse);
  rpc DownloadFile(DownloadFileRequest) returns (stream DownloadFileResponse);
}

message CreateFileRequest {
  string filename = 1;
  string content = 2;
}

message CreateFileResponse {
  bool success = 1;
  string message = 2;
}

message ReadFileRequest {
  string filename = 1;
}

message ReadFileResponse {
  bool success = 1;
  string content = 2;
  string message = 3;
}

message WriteFileRequest {
  string filename = 1;
  string content = 2;
  bool append = 3;
}

message WriteFileResponse {
  bool success = 1;
  string message = 2;
}

message DeleteFileRequest {
  string filename = 1;
}

message DeleteFileResponse {
  bool success = 1;
  string message = 2;
}

message ListFilesRequest {
  string directory = 1;
}

message ListFilesResponse {
  repeated string files = 1;
  repeated string directories = 2;
  bool success = 3;
  string message = 4;
}

message CreateDirectoryRequest {
  string directory = 1;
}

message CreateDirectoryResponse {
  bool success = 1;
  string message = 2;
}

message GetFileInfoRequest {
  string filename = 1;
}

message FileInfo {
  string filename = 1;
  int64 size = 2;
  int64 modified_time = 3;
  bool is_directory = 4;
  string permissions = 5;
}

message GetFileInfoResponse {
  bool success = 1;
  FileInfo file_info = 2;
  string message = 3;
}

message UploadFileRequest {
  oneof data {
    FileMetadata metadata = 1;
    bytes chunk = 2;
  }
}

message FileMetadata {
  string filename = 1;
  int64 file_size = 2;
}

message UploadFileResponse {
  bool success = 1;
  string message = 2;
}

message DownloadFileRequest {
  string filename = 1;
}

message DownloadFileResponse {
  oneof data {
    FileMetadata metadata = 1;
    bytes chunk = 2;
  }
}
